### PR TITLE
[RFR] Remove duplicate API call in ReferenceArrayInput

### PR DESCRIPTION
**Pull request by fargito**
Fixes https://github.com/marmelab/react-admin/issues/3108

### CODE REVIEW

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx [hunk 1]
```diff
@@ -232,7 +233,8 @@ export class UnconnectedReferenceArrayInputController extends Component<
                     'The value of ReferenceArrayInput should be an array'
                 );
             }
-            crudGetMany(reference, ids);
+            const idsToFetch = isInitialCall ? ids : ids.filter(id => !this.props.input.value.includes(id))
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Please use english :wink

[the remark this answers is not in the document]
Comment (fargito) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: > Ca marche https://lodash.com/docs/4.17.11#difference ?

Wouldn't https://lodash.com/docs/4.17.11#difference work better ?

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx [hunk 2]
```diff
@@ -4,6 +4,7 @@ import debounce from 'lodash/debounce';
 import compose from 'recompose/compose';
 import { createSelector } from 'reselect';
 import isEqual from 'lodash/isEqual';
+import difference from 'lodash/difference'
```
Comment (khoanguyenxuan) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Should we use the comma at the end of line style?

Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Yes, please configure your editor to apply our prettier rules

Comment (fargito) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Ah yes my bad. @djhi I noticed that `make prettier` changed a whole bunch of files that don't seem to respect those rules. Would it be ok to create a separate pull request with all those changes ?

Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Please don't. Can you just apply prettier on the file you modified? If you can automate it and our make command change too many files, just run the `lint` make command and fix manually. Thanks!

**Comment by fargito** [comment]
@djhi is it ok for you ? :)

**Comment by djhi** [comment]
Looks good but I won't have time to test it before next week. I'm at react europe

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx [hunk 3]
```diff
@@ -164,7 +165,8 @@ export class UnconnectedReferenceArrayInputController extends Component<
     }
 
     componentDidMount() {
-        this.fetchReferencesAndOptions(this.props);
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Why did you change that?

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx [hunk 4]
```diff
@@ -232,7 +234,10 @@ export class UnconnectedReferenceArrayInputController extends Component<
                     'The value of ReferenceArrayInput should be an array'
                 );
             }
-            crudGetMany(reference, ids);
+            const idsToFetch = isInitialCall
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Why dod you need the isInitialCall? The difference should work in all cases.

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 5]
```diff
@@ -417,6 +417,44 @@ describe('<ReferenceArrayInputController />', () => {
...
+        expect(
+            crudGetMany.mock.calls[crudGetMany.mock.calls.length - 1]
+        ).toEqual([defaultProps.reference, [5]]);
+        wrapper.setProps({ input: { value: [5, 6] } });
+        expect(
+            crudGetMany.mock.calls[crudGetMany.mock.calls.length - 1]
+        ).toEqual([defaultProps.reference, [6]]);
+    });
+
+    it('should call crudGetMany with empty list when already fetched input value changes', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: No, it should not call crudGetMany at all in that case.

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx [hunk 6]
```diff
@@ -232,7 +234,10 @@ export class UnconnectedReferenceArrayInputController extends Component<
...
+                ? ids
+                : difference(ids, this.props.input.value);
+            crudGetMany(reference, idsToFetch);
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: if `idsToFetch` is empty, the action shouldn't even be dispatched.

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx (hunk 5 above)
Comment (fargito) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: Ah ok, that's what I did at first but it broke previous tests. I will edit this

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx (hunk 4 above)
Comment (fargito) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Actually not, when the component just mounted, if we just call `fetchReferences(this.props)`, the difference between the `props` argument of the `fetchReferences` function and `this.props` results in an empty list (they're the same object of course :D). However we still need to call `crudGetMany` with all the values I think

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx (hunk 6 above)
Comment (fargito) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Yes, I will change this :)

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx (hunk 4 above)
Comment (fargito) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: I edited this, I think it should be cleaner now

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx (hunk 3 above)
Comment (fargito) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: Edited

**Comment by fzaninotto** [comment]
Thanks!
